### PR TITLE
160 Add Thread Interdependency Test for ChronoLog Client API

### LINKED ISSUE #160: Implement Thread Interdependency Test for Enhanced API Call Accuracy
### Issue Overview
Create a thread interdependency test as part of the ChronoLog correctness testing pipeline. This test will involve shared memory for all threads to achieve a more accurate global picture and ensure a determined output from API calls.

### Objectives
- Shared State Array: Enhance existing client API tests (e.g., client_multi_pthread, client_multi_argobots) by introducing a shared array of states between threads.
- Interdependent Operations: Ensure that the return code for one thread operation correctly depends on the actions of other threads, improving the accuracy and reliability of the tests.

### PR DESCRIPTION
**Pull request by iameneko**
### Summary:

This PR introduces a thread interdependency test to the ChronoLog correctness testing pipeline. The test ensures that threads perform operations in coordination, relying on shared memory to achieve a consistent global state. The test improves the accuracy of multi-threaded API interactions, ensuring a determined outcome based on the interdependencies between threads.

### Key Changes:

- **Shared State Array**: A shared array for thread states is introduced, protected by a mutex and a condition variable to ensure proper synchronization.
- **Thread State Bitmasking**: Defined bitmask states for each thread operation (initialization, chronicle creation, story acquisition, logging, and destruction).
- **State Updates and Synchronization**: All threads now update the shared state array and wait for all threads to reach specific states before proceeding, ensuring proper order of operations.

### Testing:

- Applied and tested with all multi-threaded client tests (`client_lib_multi_storytellers`, `client_lib_multi_pthread_test`, `client_lib_multi_argobots_test`, `client_lib_hybrid_argobots_test` & `client_lib_multi_openmp_test`) ensuring interdependencies between threads are handled correctly.
- Verified that operations such as chronicle creation, story acquisition, logging, and destruction are synchronized and error-free.

### CODE REVIEW

File: test/integration/Client/client_lib_multi_storytellers.cpp [hunk 1]
```diff
@@ -16,80 +19,129 @@ struct thread_arg
 
 chronolog::Client*client;
 
+std::vector <int> shared_state;
```
Comment (fkengun) [inline test/integration/Client/client_lib_multi_storytellers.cpp]: These shared global variables, definitions, and functions are better to be put in a header file and included in each test case.

File: test/integration/Client/client_lib_multi_storytellers.cpp [hunk 2]
```diff
@@ -16,80 +19,129 @@ struct thread_arg
...
+std::mutex state_mutex;
+std::condition_variable state_cv;
+
+// Define bitmask for thread states
+enum ThreadState
+{
+    THREAD_INITIALIZED = 1 << 0, // 00000001 (binary), 1 (decimal)
+    CHRONICLE_CREATED = 1 << 1, // 00000010 (binary), 2 (decimal)
+    STORY_ACQUIRED = 1 << 2, // 00000100 (binary), 4 (decimal)
+    FINISHED_LOGGING = 1 << 3, // 00001000 (binary), 8 (decimal)
```
Comment (fkengun) [inline test/integration/Client/client_lib_multi_storytellers.cpp]: Better to say `DATA_ACCESS_FINISHED`. Logging could be ambiguous. And we will have Read-after-Write or Write-after-Read test cases in the future.

File: test/integration/Client/client_lib_hybrid_argobots_test.cpp [hunk 3]
```diff
@@ -10,14 +10,50 @@
 
 chronolog::Client*client;
 
+// Define bitmask for thread states
+enum ThreadState
+{
```
Comment (ibrodkin) [inline test/integration/Client/client_lib_hybrid_argobots_test.cpp]: It's a good practice to leave 0000 for the "zero" state , I would call it "UNKNOWN"

File: test/integration/Client/client_lib_hybrid_argobots_test.cpp [hunk 4]
```diff
@@ -10,14 +10,50 @@
...
+std::condition_variable state_cv;
+
 struct thread_arg
 {
     int tid;
 };
 
+// Function to update shared state and wait for all threads to reach a specific state
+void update_shared_state_and_wait(int tid, int target_state_bitmask)
+{
```
Comment (ibrodkin) [inline test/integration/Client/client_lib_hybrid_argobots_test.cpp]: I suggest that each thread updates only its own state. This way we know that the only writer on the shared_state[i] is thread i , every other thread has read only access to shared_state[i] making the mutex unnecessary.

File: test/integration/Client/client_lib_multi_argobots_test.cpp [hunk 5]
```diff
@@ -39,43 +77,44 @@ void thread_function(void*tt)
               , chronicle_name, ret);
```
Comment (ibrodkin) [inline test/integration/Client/client_lib_multi_argobots_test.cpp]: Please, make this paragraph more nuanced based on the state of the other threads.

**Review by ibrodkin — CHANGES_REQUESTED** [review]
1. The idea of having the array of thread states is about tracking each thread state individually.  I suggest that each thread updates only its own state. This way we know that the only writer on the _shared_state[i]_ element of the array is thread _i_ , every other thread has read only access to _shared_state[i]_ making the mutex unnecessary. 
Please, remove the synchronization mutex.
2. Please use bitmask operations on the bitmask values, this allows for more nuanced state tracking.
3. Please, make use of the sshared state array values to make validation of the return values coming from the visor more nuanced.

**Pull request closed without merging**